packages/swing-store/src/transcriptStore.js

```diff
@@ -0,0 +1,383 @@
+    assert(bounds, `no current transcript for ${vatID}`);
+    return bounds;
+  }
+
+  const sqlEndCurrentTranscript = db.prepare(`
+    UPDATE transcriptSpans
+    SET current = 0
+    WHERE current = 1 AND vatID = ?
+  `);
+
+  function rolloverSpan(vatID) {
+    ensureTxn();
+    const { hash, startPos, endPos, size } = getCurrentTranscriptBounds(vatID);
+    noteExport(
+      artifactName({ vatID, startPos }),
+      JSON.stringify({ vatID, hash, startPos, endPos, size }),
+    );
+    sqlEndCurrentTranscript.run(vatID);
+    sqlWriteSpan.run(vatID, endPos, endPos, initialHash, 1, 0);
+  }
+
+  function* getArtifactNames(includeHistorical) {
+    const sql = includeHistorical
+      ? sqlGetAllTranscriptMetadata
+      : sqlGetCurrentTranscriptMetadata;
+    for (const rec of sql.iterate()) {
+      yield artifactName(rec);
+    }
+  }
+
+  const sqlGetTranscriptEndPos = db.prepare(`
+    SELECT endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos = ?
+  `);
+  sqlGetTranscriptEndPos.pluck(true);
+
+  const sqlReadTranscriptItems = db.prepare(`
+    SELECT item
+    FROM transcriptItems
+    WHERE vatID = ? AND ? <= position AND position < ?
+    ORDER BY position
+  `);
+
+  /**
+   * @param {string} vatID
+   * @param {number} [startPos]
+   */
+  function readTranscript(vatID, startPos) {
+    let endPos;
+    if (startPos === undefined) {
+      ({ startPos, endPos } = getCurrentTranscriptBounds(vatID));
+    } else {
+      assert.typeof(startPos, 'number');
+      endPos = sqlGetTranscriptEndPos.get(vatID, startPos);
+      assert.typeof(
+        endPos,
+        'number',
+        `no transcript for ${vatID} at ${startPos}`,
+      );
+    }
+    !transcriptStatus.get(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+    insistTranscriptPosition(startPos);
+    startPos <= endPos || Fail`${q(startPos)} <= ${q(endPos)}}`;
+
+    function* reader() {
+      for (const { item } of sqlReadTranscriptItems.iterate(
+        vatID,
+        startPos,
+        endPos,
+      )) {
+        transcriptStatus.get(vatID) === 'reading' ||
+          Fail`can't read transcript ${q(vatID)}, it's been closed`;
+        yield item.substring(0, item.length - 1);
+      }
+      transcriptStatus.delete(vatID);
+    }
+    !transcriptStatus.has(vatID) ||
+      Fail`can't read transcript ${q(vatID)} because it's already in use`;
+
+    if (startPos === endPos) {
+      return empty();
+    }
+
+    transcriptStatus.set(vatID, 'reading');
+
+    return reader();
+  }
+
+  const nl = Buffer.from('\n');
+
+  /**
+   * @param {string} vatID
+   * @param {number} startPos
+   * @returns {AsyncIterable<Uint8Array>}
+   */
+  function getTranscript(vatID, startPos) {
+    async function* fetcher() {
+      for (const entry of readTranscript(vatID, startPos)) {
+        const item = typeof entry === 'string' ? Buffer.from(entry) : entry;
+        yield item;
+        yield nl;
```

Comment:
I'm feeling weirdly conflicted here. On one end concatenations are expensive if not needed, on the other reifying the delimiter as a separately yielded value has sometimes weird impact on downstream consumers (for example it causes trashing in most readline transforms). In modern engines I think I'd prefer to do the following:

```suggestion
        yield Buffer.from(`${entry}\n`);
```

But then I'm even more confused because I thought transcript entries already contained a terminating `\n`?

Reply:
Yeah, this has driven me crazy and the implementation has oscillated quite a bit before settling on its current form.  The best I can say about the current form is that the implementation has at least been hammered into being self-consistent in its behavior.  The problem is that transcript spans need to delimit items but none of the code that consumes or produces the items actually cares whether the terminating newline is there or not.  Since the transcript spans are transmitted in a stream the terminator is necessary in case the chunks get repacked by some intermediary.  Because the underlying database actually stores the items as individual records, *it* does not need the terminator, but we append the terminator (if the item string doesn't already have one) before storing the item so that we can spit the items out without processing.  However, that means that the reader needs to strip the terminator off before handing the item back so that the database reads and writes will round-trip, so there's this awkward dance of various entities adding or removing the terminator as the item flows through processing   One key motivator for the currently implemented scheme was to be able to compute the hash of the span artifact from the raw stream without having to understand its internal structure, but now we don't actually do that since hashes are only expected to be checked at the receiving endpoint. However, the importer checks a span by computing a running hash on an item-by-item basis.  I think this means we could switch to only using the newlines while the span is flowing through the export stream and get rid of the add a newline/remove a newline dance.  Probably we should do this.